### PR TITLE
adapter: implement `EXPLAIN OPTIMIZER TRACE`

### PR DESCRIPTION
**Pull request by aalexandrov**
This basically is everything that we need to close the #13139 epic.

### Motivation

The optimizer team (and in the future other orgs, such as field engineering and customer support) might need fined access to fine-grained representations about the query compilation lifecycle to troubleshoot problematic queries and advise customers.

Current capabilities provide such insight for four distinct phases in [the compiler pipeline](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/101-query-compilation.md), but this is not sufficient when troubleshooting issues in specific transforms or arising due to friction between transforms.

This PR implements a requested feature (#13139). More specifically, it implements handling of the

```sql
EXPLAIN OPTIMIZER TRACE AS $FORMAT ...
```

command. The output is not very useful on its own, but with the right UI it can be a powerful debugging tool.

To try out this feature, please use [the complementary UI](https://psychic-winner-f678af23.pages.github.io/) (UI sources are in https://github.com/MaterializeInc/mztrace-explorer). The PR has attached [an archive with the outputs of all TPC-H queries](https://github.com/MaterializeInc/materialize/files/9618809/tpch.zip) for you to explore. Here is a screenshot for a random stage from Q03:

Screenshot from 2022-09-21 19-52-43

### Tips for reviewer

I will add some insights about the code structure later today or tomorrow. For now, I just want to point out @guswynn is best-positioned to review 

- the second commit and the `optimizer_trace` additions in the third commit.
- the compute team is best-positioned to review the instrumentation of the optimizer pipeline and the `sequence_explain_new` rewrite in the last two commits.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - `EXPLAIN OPTIMIZER TRACE` is not publicly advertised yet.

### CODE REVIEW

**Comment by ggevay** [comment]
This looks awesome! It will be good not only for debugging, but also to learn about the optimizer.

I'm trying to try it out, but getting an error after uploading the trace:
```
Oh snap! You got an error!
SyntaxError: Unexpected token 'T', " Time | "... is not valid JSON
```
I guess I'm collecting the trace in a wrong way somehow. I did
```
explain optimizer trace as json for SELECT * FROM t2;
```
The psql console then shows
```
  Time   |                                       Path                                        |                     Plan                      
---------+-----------------------------------------------------------------------------------+-----------------------------------------------
 8574    | optimize/raw                                                                      | {                                            +
         |                                                                                   |   "Get": {                                   +
         |                                                                                   |     "id": {                                  +
         |                                                                                   |       "Global": {                            +
         |                                                                                   |         "User": 4                            +
         |                                                                                   |       }                                      +
         |                                                                                   |     },                                       +
         |                                                                                   |     "typ": {                                 +
         |                                                                                   |       "column_types": [                      +
         |                                                                                   |         {                                    +
         |                                                                                   |           "scalar_type": "Float64",          +
         |                                                                                   |           "nullable": true                   +
         |                                                                                   |         },                                   +
         |                                                                                   |         {                                    +
         |                                                                                   |           "scalar_type": "Float64",          +
         |                                                                                   |           "nullable": false                  +
         |                                                                                   |         }                                    +
         |                                                                                   |       ],                                     +
         |                                                                                   |       "keys": []                             +
         |                                                                                   |     }                                        +
         |                                                                                   |   }                                          +
         |                                                                                   | }
 120357  | optimize/hir_to_mir                                                               | {                                            +
         |                                                                                   |   "Let": {                                   +
         |                                                                                   |     "id": 0,                                 +
         |                                                                                   |     "value": {                               +
         |                                                                                   |       "Constant": {                          +
         |                                                                                   |         "rows": {                            +
         |                                                                                   |           "Ok": [                            +
         |                                                                                   |             [                                +
         |                                                                                   |               {                              +
         |                                                                                   |                 "data": []                   +
         |                                                                                   |               },                             +
         |                                                                                   |               1                              +
         |                                                                                   |             ]                                +
         |                                                                                   |           ]                                  +
         |                                                                                   |         },                                   +
         |                                                                                   |         "typ": {                             +
         |                                                                                   |           "column_types": [],                +
         |                                                                                   |           "keys": []                         +
         |                                                                                   |         }                                    +
         |                                                                                   |       }                                      +
         |                                                                                   |     },                                       +
         |                                                                                   |     "body": {                                +
         |                                                                                   |       "Join": {                              +
         |                                                                                   |         "inputs": [                          +
         |                                                                                   |           {                                  +
         |                                                                                   |             "Get": {                         +
         |                                                                                   |               "id": {                        +
         |                                                                                   |                 "Local": 0                   +
         |                                                                                   |               },                             +
         |                                                                                   |               "typ": {                       +
         |                                                                                   |                 "column_types": [],          +
         |                                                                                   |                 "keys": [                    +
         |                                                                                   |                   []                         +
         |                                                                                   |                 ]                            +
```
(and many more lines)

It shows this in `less`, from which I wrote it out into a file by hitting `s`, and then uploaded the file, which resulted in the above error.

**Comment by aalexandrov** [comment]
@ggevay the output needs to be post-processed as JSON. The latest version of https://github.com/aalexandrov/mzt has a new command `trace` that was used for the tpch queries, but I'm trying to figure out whether I can add a query prompt directly in the UI.

**Review by ggevay — COMMENTED** [review]
Just a few minor comments for now, I'll continue tomorrow.

I'm wondering if it would be possible to add tracing for `MirScalarExpr::reduce`. I guess it would create the issue that all the transforms that call it would become blue, because it would look like as if the plan changed, but actually we would be just printing a part of the plan. I'm not sure where is the mechanism that decides whether to make something blue, but maybe we could teach it that it should skip those events that come from `MirScalarExpr::reduce`, or that it should track changes there separately.

File: src/ore/src/tracing/optimizer.rs
Comment (teskje) [reply src/ore/src/tracing/optimizer.rs:191]: You need to enable the `tracing_` feature:
```
cargo test --features tracing_ -- test_optimizer_trace
```

**Comment by aalexandrov** [comment]
The result of the `EXPLAIN OPTIMIZER TRACE` needs to be post-processed into a `*.json` format expected by the UI. Currently, the only semi-automated way to do this is to use the latest mzt version:
```sh
mzt trace query "$(cat $QUERY.sql)" > trace.json
```
For example, for the TPC-H queries in the PR I used the following command:
```sh
for i in {01..22}; do
    mzt trace query --db-name tpch \
      "$(cat ~/benchmarks/tpch/workload/Q${i}.sql)" >    
      "~/benchmarks/tpch/optimizer/Q${i}.json";
done
```
I'm trying to figure out what it will take to extend the UI with some ad-hoc querying capabilities, but this is still WIP.

**Comment by aalexandrov** [comment]
> I'm wondering if it would be possible to add tracing for `MirScalarExpr::reduce`.

I'm not sure that this is a good fir in the current UI framework (especially as a default). Can you tell me use-cases where this would be useful? At the moment, you can see the end result of applying reduce as part of the surrounding MIR transform.

If its primarily for debugging code you can always sprinkle more `trace_plan` code in your code while you are developing. The enclosing function will have to be instrumented with `#[tracing::instrument]` and there should be exactly one plan for every scope that represents the final result for the UI to work reasonably with your instrumentation (I probably should document it in the `lib` modules).

If you are worried about having to debug such issues directly from a customer trace, I suggest to open a follow-up issue where we discuss the options. One idea is to rename `EXPLAIN OPTIMIZER TRACE` to `EXPLAIN OPTIMIZER PIPELINE` and control the output level `(detail_log, detail_debug, detail_trace)` as an additional `WITH()` parameter. Really fine-grained details like the iterations of an `MirScalarExpr::reduce` will be then represented only at the lowest level, which we would be able to request from customers if needed.

File: src/adapter/src/explain_new/optimizer_trace.rs
Comment (teskje) [inline src/adapter/src/explain_new/optimizer_trace.rs:92]: I think `sort_by_key` can be used here and would be more efficient:

```
results.sort_by_key(|x| x.instant);
```

... or `sort_unstable_by_key`, if unstable sorting is fine.

**Comment by ggevay** [comment]
>> I'm wondering if it would be possible to add tracing for MirScalarExpr::reduce.

> I'm not sure that this is a good fit in the current UI framework (especially as a default). Can you tell me use-cases where this would be useful?

I had a feeling that there might be something going on inside `reduce` in https://github.com/MaterializeInc/materialize/issues/14905, but I'm not sure. I'll try to actually apply the tracing to that situation today, and we'll see. Thanks for posting the `mzt` commands!

**Review by teskje — APPROVED** [review]
This looks great! Thanks for splitting the PR into sensible commits too!

File: src/transform/src/fusion/join.rs
Comment (aalexandrov) [reply src/transform/src/fusion/join.rs:41]: Done.

File: src/ore/src/tracing/optimizer.rs
Comment (aalexandrov) [reply src/ore/src/tracing/optimizer.rs:237]: Argh, I had to find that answer the hard way myself.

File: src/adapter/src/explain_new/optimizer_trace.rs
Comment (aalexandrov) [reply src/adapter/src/explain_new/optimizer_trace.rs:54]: Done.

File: src/ore/src/tracing/optimizer.rs
Comment (aalexandrov) [reply src/ore/src/tracing/optimizer.rs:156]: I renamed this method to `drain_as_vec`.

**Comment by aalexandrov** [comment]
> I'll try to actually apply the tracing to that situation today, and we'll see. Thanks for posting the `mzt` commands!

I tweaked `OptimizerTrace` so this works out of the box. This should work fine once you cherry-pick [this commit](https://github.com/aalexandrov/materialize/commit/8fe049baf9f8cd8be193f65e922865587e7a6761) on top of the HEAD of this PR to see the `reduce` iteration states (under `*/reduce_mir_scalar`

**Comment by ggevay** [comment]
A use case for tracing details of individual transforms: In [this trace](https://gist.github.com/ggevay/b8c27ceff2400a6cb8eae87f7edcd2d1), take a look at the blue `predicate_pushdown`. There is _a lot_ of magic going on in that single transform: First of all, it does many steps, and also it’s tricky that when it pushes the predicate through the `Map (null, null)`, it inlines the `nulls` to create stuff like `xxx = null`, and then later (but still inside `predicate_pushdown`) `MirScalarExpr::reduce` simplifies these to just `null`. I’m still figuring out the details, but I’m thinking this `reduce` might be missing some other opportunity after the variadic commit.

But anyhow, adding more detailed tracing can be follow-up work. This PR looks good as it is. It already helped debug the above issue, by restricting my attention to `predicate_pushdown`. If I hadn’t seen that `predicate_pushdown` introduced all those null stuff, I would have probably been looking for what could introduce those null stuff in some other transforms.

**Review by ggevay — APPROVED** [review]

**Comment by ggevay** [comment]
The problem was indeed inside `reduce`, see https://github.com/MaterializeInc/materialize/issues/14905#issuecomment-1256621236

**Comment by aalexandrov** [comment]
Here is the result of `bin/optbench compare` command that compares the current `main` (`a330eb674`) vs the head of this PR (`97864e016`)

```text
~/workspace/materialize explain_trace ❯ bin/optbench compare ~/benchmarks/tpch/optimize/optbench-tpch-v0.27.0-alpha.22.post-a330eb674.csv ~/benchmarks/tpch/optimize/optbench-tpch-v0.27.0-alpha.22.post-97864e016.csv 
Compare experiment results between "/home/alexander/benchmarks/tpch/optimize/optbench-tpch-v0.27.0-alpha.22.post-a330eb674.csv" and "/home/alexander/benchmarks/tpch/optimize/optbench-tpch-v0.27.0-alpha.22.post-97864e016.csv"
base times
----------
              Q01         Q02        Q03        Q04        Q05        Q06        Q07        Q08        Q09        Q10        Q11        Q12        Q13        Q14        Q15        Q16        Q17        Q18         Q19         Q20         Q21         Q22
amin     466000.0   5367000.0   946000.0  1752000.0  1721000.0   714000.0  2917000.0  3143000.0  1799000.0  1263000.0  2476000.0   992000.0  2042000.0  1347000.0  1561000.0  2987000.0  3992000.0  2612000.0   6624000.0   6560000.0   6107000.0   5746000.0
median   971000.0   8735000.0   979000.0  1806000.0  1763000.0   734000.0  2984000.0  3231000.0  1847000.0  1295500.0  2543500.0  1017000.0  2103000.0  1379000.0  1602000.0  3053500.0  4086000.0  2679000.0   6784000.0   6723500.0   6258000.0   5904000.0
amax    1562000.0  11219000.0  2242000.0  5040000.0  4061000.0  1764000.0  6826000.0  7371000.0  4145000.0  3281000.0  5930000.0  2351000.0  4743000.0  3069000.0  3786000.0  7502000.0  9251000.0  6361000.0  14406000.0  13819000.0  11893000.0  11727000.0

diff times
----------
              Q01         Q02        Q03        Q04        Q05        Q06        Q07        Q08        Q09        Q10        Q11        Q12        Q13        Q14        Q15        Q16        Q17        Q18         Q19         Q20         Q21         Q22
amin     476000.0   5465000.0   959000.0  1780000.0  1745000.0   730000.0  2948000.0  3206000.0  1829000.0  1280000.0  2514000.0  1011000.0  2085000.0  1369000.0  1587000.0  3023000.0  4024000.0  2655000.0   6731000.0   6668000.0   6201000.0   5886000.0
median  1004500.0   8614500.0   994000.0  1827500.0  1792000.0   749500.0  3013000.0  3285000.0  1875000.0  1314500.0  2582000.0  1033000.0  2132000.0  1401000.0  1631500.0  3102500.0  4126000.0  2716000.0   6874000.0   6801500.0   6331000.0   6000500.0
amax    1319000.0  11341000.0  2435000.0  4218000.0  4276000.0  1751000.0  6476000.0  7146000.0  4109000.0  2800000.0  5853000.0  2450000.0  5492000.0  3429000.0  4002000.0  6870000.0  9567000.0  6803000.0  12835000.0  12168000.0  12278000.0  11133000.0

diff/base ratio
---------------
             Avg       Q01       Q02       Q03       Q04       Q05       Q06       Q07       Q08       Q09       Q10       Q11       Q12       Q13       Q14       Q15       Q16       Q17       Q18       Q19       Q20       Q21       Q22
amin    1.016548  1.021459  1.018260  1.013742  1.015982  1.013945  1.022409  1.010627  1.020045  1.016676  1.013460  1.015347  1.019153  1.021058  1.016333  1.016656  1.012052  1.008016  1.016462  1.016153  1.016463  1.015392  1.024365
median  1.014241  1.034501  0.986205  1.015322  1.011905  1.016449  1.021117  1.009718  1.016713  1.015160  1.014666  1.015137  1.015733  1.013790  1.015954  1.018414  1.016047  1.009790  1.013811  1.013267  1.011601  1.011665  1.016345
amax    0.987308  0.844430  1.010874  1.086084  0.836905  1.052943  0.992630  0.948725  0.969475  0.991315  0.853398  0.987015  1.042110  1.157917  1.117302  1.057052  0.915756  1.034158  1.069486  0.890948  0.880527  1.032372  0.949348
```

Because `bin/optbench` uses the old explain format, this is effectively also benchmarking optimization runs without active instrumentation. The number suggest that in the absence of an the tracing subscriber stack wrapped by `OptimizerTrace` the net effect of all `trace_plan` calls is 1% on average (see the `median` in the `diff/base ratio` column).

File: src/adapter/src/explain_new/optimizer_trace.rs
Comment (aalexandrov) [inline src/adapter/src/explain_new/optimizer_trace.rs:37]: Because we have a lot of different plan types `T` the only way to collect the trace at the moment is to enumerate all of them as a chain of layered `PlanTrace<T>` listeners, one for each plan.

Once `Valuable` is stabilized I can probably do this in a more efficient way.

File: src/ore/src/tracing/optimizer.rs
Comment (aalexandrov) [inline src/ore/src/tracing/optimizer.rs:92]: Conceptually, this should just attach an event of type `T` to the current span. However, since typed events are not yet supported, I implemented this a bit differently. Hopefully the implementation can be simplified once `Valuable` is stable.

File: src/ore/src/tracing/optimizer.rs
Comment (aalexandrov) [inline src/ore/src/tracing/optimizer.rs:24]: Note that the `path` and `times` stack accumulated in the `Layer` implementation for `PlanTrace<T>` can be shared by a chain of `PlanTrace<T>` instances wrapped in a layered subscriber, but at the moment this work is replicated.

I didn't find a clean way to do this kind of bookkeeping in the `Registry` base of the layer chain.

File: src/ore/src/tracing/optimizer.rs
Comment (aalexandrov) [inline src/ore/src/tracing/optimizer.rs:94]: Performance of `trace_plan` instrumentation is based on the assumption that this `downcast_ref` call is not going to be very expensive if the layer does not contain a `PlanTrace<T>` instance.

File: src/ore/src/tracing/optimizer.rs
Comment (aalexandrov) [inline src/ore/src/tracing/optimizer.rs:243]: nit: If the `Dispatch` API had a `downcast_mut` counterpart I would be able to define `drain_as_vec` on `&mut self`.

**Review by aalexandrov — COMMENTED** [review]
@guswynn I added some comments with remarks on some non-polished aspects of the `OptimizerTrace` and `mz_ore::tracing::optimizer` implementations.

File: src/adapter/src/coord/sequencer.rs
Comment (guswynn) [inline src/adapter/src/coord/sequencer.rs:2262]: i like this, the `raw` and `optimize` next to each other!

I'm not an expert so i don't know what `local` means, but the tracing components of this change make sense!

File: src/adapter/src/explain_new/optimizer_trace.rs
Comment (guswynn) [reply src/adapter/src/explain_new/optimizer_trace.rs:37]: this seems reasonable for now as a code organization technique, instead of making `PlanTrace` have to match on all them!

File: src/adapter/src/explain_new/optimizer_trace.rs
Comment (guswynn) [inline src/adapter/src/explain_new/optimizer_trace.rs:179]: in what case do we actually hit this case? should we just panic? If we are draining entries, then its a logic bug in our code somewhere to have not entered the `OptimizerTrace` dispatch right?

Also i think what you have done here is reasonable, to just downcast the `Dispatch` itself to collect all the trace info into the final output!

File: src/ore/src/tracing/optimizer.rs
Comment (guswynn) [reply src/ore/src/tracing/optimizer.rs:92]: I actually like what you have done here with the `PlanTrace<T>` layers, but I am curious, once we have valuable, can we remove this function, and just have something like `tracing::trace!(plan.as_value())`? Needing to do this `with_subscriber` stuff makes sense for how you implemented it, but avoiding that in the future seems cleaner

File: src/ore/src/tracing/optimizer.rs
Comment (guswynn) [inline src/ore/src/tracing/optimizer.rs:1]: I don't think `ore` is the right place for all this, it should probably be in some other crate.

File: src/ore/src/tracing/optimizer.rs
Comment (guswynn) [reply src/ore/src/tracing/optimizer.rs:24]: is `PathBuf` the right type here? or is just convenient based on how you implemented it? `String` might make more sense, as we aren't working with unix file paths

File: src/ore/src/tracing/optimizer.rs
Comment (guswynn) [inline src/ore/src/tracing/optimizer.rs:116]: how much slower is it when you have tracing on? One big change we could make here is using span `extensions_mut` to store the data instead: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/registry/trait.SpanData.html#tymethod.extensions_mut

I am not going to block this pr for this change, but i think it'd be worth it in the future. However, you might find that using a `RwLock` instead of a `Mutex` might be a perf improvement, depending on how often you are reading. (underneath, `ExtensionsMut` is using a `RwLock`)

File: src/ore/src/tracing/optimizer.rs
Comment (guswynn) [inline src/ore/src/tracing/optimizer.rs:119]: a comment explaining this default behavior would be ncie

File: src/ore/src/tracing/optimizer.rs
Comment (guswynn) [inline src/ore/src/tracing/optimizer.rs:135]: is there only one time stack for the whole optimization (well, per-`T`)

File: src/ore/src/tracing/optimizer.rs
Comment (guswynn) [reply src/ore/src/tracing/optimizer.rs:243]: This will probably never be added so doing the coordination yourself is fine!

File: src/ore/src/tracing/optimizer.rs
Comment (guswynn) [inline src/ore/src/tracing/optimizer.rs:312]: love that you are unit-testing this

File: src/ore/src/tracing/optimizer.rs
Comment (guswynn) [inline src/ore/src/tracing/optimizer.rs:215]: this is great haha!

File: src/ore/src/tracing/optimizer.rs
Comment (guswynn) [reply src/ore/src/tracing/optimizer.rs:135]: Also, a comment explaining how this works would be nice, when the span is exited (not closed), we pop from the path stack; what actually reads from the path stack?

Comment (guswynn) [reply src/ore/src/tracing/optimizer.rs:135]: Oh I see, its `current_path`

I believe this will all be cleaned up if we use `extensions_mut` instead

File: src/ore/src/tracing/optimizer.rs
Comment (guswynn) [inline src/ore/src/tracing/optimizer.rs:173]: this kindof assume that each span only has 1 child right? or am i mis-reading

This requirement is fine, but we should either document it, or panic if we can detect it being violated

**Review by guswynn — APPROVED** [review]
My comments are mostly questions and improvements, and should not block merging this! My accept is related to the tracing stuff, as I have less context on the actual optimizer stuff

The three improvements that I think should happen are:
- Store the trace data on the span's data extensions itself, and only doing the coarse, mutex locking `on_close`, when the span is dropped
- move the logic into its own crate, out of `mz-ore`
- more comments on how the timing and time stack work

File: src/ore/src/tracing/optimizer.rs
Comment (aalexandrov) [reply src/ore/src/tracing/optimizer.rs:24]: I didn't realize that `PathBuf` is meant to abstract over OS paths and has `os` stuff inside. I will change this to `Vec<String>`.

File: src/adapter/src/explain_new/optimizer_trace.rs
Comment (aalexandrov) [reply src/adapter/src/explain_new/optimizer_trace.rs:179]: Yes, we definitely can panic here because these are private methods.

File: src/ore/src/tracing/optimizer.rs
Comment (aalexandrov) [reply src/ore/src/tracing/optimizer.rs:173]: Each span can have only one plan attached to it as an invariant and this plan should represent the end state of that span. This is documented as constraint (2) in the `trace_plan` function.

Spans can be nested in each other in a tree-like manner as with regular `tracing`.

File: src/ore/src/tracing/optimizer.rs
Comment (aalexandrov) [reply src/ore/src/tracing/optimizer.rs:135]: > is there only one time stack for the whole optimization

Yes, there is one time and one path stack conceptually, and for lack of further optimizing this at the moment we are going to maintain one copy per plan type `T`.

**Comment by aalexandrov** [comment]
> My comments are mostly questions and improvements, and should not block merging this! My accept is related to the tracing stuff, as I have less context on the actual optimizer stuff
> 
> The three improvements that I think should happen are:
> 
> * Store the trace data on the span's data extensions itself, and only doing the coarse, mutex locking `on_close`, when the span is dropped

I tried doing this, but I got stuck. In general I want this method

```rust
pub fn trace_plan<T: Clone + 'static>(plan: &T) {
    tracing::Span::current().with_subscriber(|(_id, subscriber)| {
        if let Some(trace) = subscriber.downcast_ref::<PlanTrace<T>>() {
            // TODO: clone `plan` and attached the cloned `plan: T` to the extensions_mut of the current scope
        }
    });
}
```

However its not clear how to implement the `TODO` in the absence of a `tracing_subscriber::layer::Context`. In a `layer::Layer` implementation, I can do as follows:

```rust
fn on_event(&self, event: &::tracing::Event<'_>, ctx: tracing_subscriber::layer::Context<'_, S>) {
    let extensions_mut = ctx.event_span(event).unwrap().extensions_mut();
}
```

but here I don't have `ctx` and I wasn't able to figure out how to derive it.

> * move the logic into its own crate, out of `mz-ore`

At the moment this will be a create consisting of only one file (`mz_ore::tracing::optimizer`), so I'm not sure if its worthwhile doing this just yet. @benesch thoughts?

> * more comments on how the timing and time stack work

I'll do this before I merge.

**Comment by guswynn** [comment]
@aalexandrov ah you are right, without valuable its not really clean/possible to switch to `extensions_mut`, we can wait for that. `valuable` is likely not too far away 

> At the moment this will be a create consisting of only one file (mz_ore::tracing::optimizer), so I'm not sure if its worthwhile doing this just yet.

Yeah this isnt blocking, code movement is easy to do later

File: src/adapter/src/explain_new/optimizer_trace.rs
Comment (guswynn) [reply src/adapter/src/explain_new/optimizer_trace.rs:179]: nice

File: src/ore/src/tracing/optimizer.rs
Comment (guswynn) [reply src/ore/src/tracing/optimizer.rs:173]: ah thats nice

**Comment by benesch** [comment]
> > At the moment this will be a create consisting of only one file (mz_ore::tracing::optimizer), so I'm not sure if its worthwhile doing this just yet.
> 
> Yeah this isnt blocking, code movement is easy to do later

Ah, yeah, please don't add stuff this specific to `mz_ore`! My rule of thumb is only things that would go in a batteries-included standard library should go in `mz_ore`. If there's some way to make the module generally useless, then it be ok, but as written it talks a lot about explain plans specifically.

A single file crate seems just fine to me—but is there nowhere existing it could go? `mz_repr::explain_new`? I'm just skimming, though, and anywhere that's not `mz_ore::tracing` works for me!

File: src/ore/src/tracing/optimizer.rs
Comment (aalexandrov) [reply src/ore/src/tracing/optimizer.rs:92]: >  can we remove this function, and just have something like `tracing::trace!(plan.as_value())?`

I would love if this is the case, but this kind of depends on the `Valuable` API. If `Valuable` can be automatically derived for types that implement `serde::ser::Serialize` this will be pretty simple, otherwise we will need to add `Valuable` implementations (possibly derived ones) for all plan types in all IR types used in the optimizer and I'm not sure it will be worth carrying this code around for the sole purpose to simplify the `PlanTrace` and `OptimizerTrace` implementations.

**Comment by aalexandrov** [comment]
@benesch: I moved this to `mz_repr::explain_new::tracing` and I'm now exporting this module behind a `tracing_` feature similar to `ore`. I'm going to auto-merge this now in order to land this in the next release. 

Let me know if you want me to reorg this further, I will open a separate PR to fix the package location and the features.

**Comment by benesch** [comment]
No, that's wonderful, thank you!

**Pull request merged**
